File: src/main/java/com/github/albertosh/flitetrakr/sync/usecases/priceofconnection/PriceOfConnectionUseCase.java
package com.github.albertosh.flitetrakr.sync.usecases.priceofconnection;

import com.google.common.base.Preconditions;

import com.github.albertosh.flitetrakr.model.Connection;
import com.github.albertosh.flitetrakr.sync.services.IConnectionService;

import java.util.List;
import java.util.Optional;

public class PriceOfConnectionUseCase
        implements IPriceOfConnectionUseCase {

    private final IConnectionService connectionService;

    public PriceOfConnectionUseCase(IConnectionService connectionService) {
        this.connectionService = Preconditions.checkNotNull(connectionService);
    }

    @Override
    public Integer execute(PriceOfConnectionUseCaseInput input) throws PriceOfConnectionUseCaseError {
        Integer accum = 0;
        List<String> codes = input.getCodes();
        int connectionSize = codes.size();
        for (int i = 1; i < connectionSize; i++) {
            Optional<Connection> optConnection = connectionService.recoverConnection(
                    codes.get(i - 1), codes.get(i)
            );
            if (optConnection.isPresent()) {
                accum += optConnection.get().getPrice();
            } else {
                throw PriceOfConnectionUseCaseError.connectionNotFound();
            }
        }
        return accum;
    }
}

File: flitetrakr_reactive_kotlin/src/main/kotlin/com/github/albertosh/flitetrakr/usecases/PriceOfConnectionUseCase.kt
package com.github.albertosh.flitetrakr.usecases

import com.github.albertosh.flitetrakr.services.ConnectionServiceError
import com.github.albertosh.flitetrakr.services.IConnectionService
import com.github.albertosh.flitetrakr.util.language.LanguageUtils
import com.github.albertosh.flitetrakr.util.language.Message
import io.reactivex.Flowable
import io.reactivex.Single
import io.reactivex.SingleSource

interface IPriceOfConnectionUseCase {

    fun execute(input: PriceOfConnectionUseCaseInput): Single<PriceOfConnectionUseCaseOutput>

}

data class PriceOfConnectionUseCaseInput(val codes: List<String>)

data class PriceOfConnectionUseCaseOutput(val price: Int)

sealed class PriceOfConnectionUseCaseError(message: String) : RuntimeException(message) {
    object connectionNotFound : PriceOfConnectionUseCaseError(
            LanguageUtils.getMessage(Message.CONNECTION_NOT_FOUND))
}


class PriceOfConnectionUseCase(
        private val service: IConnectionService)
: IPriceOfConnectionUseCase {

    override fun execute(input: PriceOfConnectionUseCaseInput): Single<PriceOfConnectionUseCaseOutput> {

        val codes = input.codes

        val pairedCities = codes
                .subList(1, codes.size)
                .mapIndexed { i, city -> Pair(codes[i], city) }

        return Flowable.fromIterable(pairedCities)
                .flatMap {
                    service.recoverConnection(it.first, it.second)
                            .toFlowable()
                }
                .map { it.price }
                .reduce { acc, price -> acc + price }
                .map { PriceOfConnectionUseCaseOutput(it) }
                .toSingle()
                .onErrorResumeNext { error ->
                    when (error) {
                        is ConnectionServiceError.connectionNotFound ->
                            SingleSource {
                                it.onError(PriceOfConnectionUseCaseError.connectionNotFound)
                            }
                        else ->
                            SingleSource {
                                it.onError(error)
                            }
                    }
                }

    }

}